test/test_Choices.py

```diff
@@ -0,0 +1,48 @@
+from test.vim_test_case import VimTestCase as _VimTest
+from test.constant import *
+
+class Choices_WillBeExpandedToInlineSelection(_VimTest):
+    snippets = ("test", "${1|red,gray|}")
+    keys = "test" + EX
+    wanted = "1.red|2.gray"
+
+class Choices_ExpectCorrectResult(_VimTest):
+    snippets = ("test", "${1|red,gray|}")
+    keys = "test" + EX + "2"
+    wanted = "gray"
+
+
+class Choices_WillAbandonSelection_If_CharTyped(_VimTest):
+    snippets = ("test", "${1|red,green|}")
+    keys = "test" + EX + "char"
+    wanted = "char"
+
+
+class Choices_WillAbandonSelection_If_InputIsGreaterThanMaxSelectionIndex(_VimTest):
+    snippets = ("test", "${1|red,green|}")
+    keys = "test" + EX + "3"
+    wanted = "3"
+
+class Choices_WilNotMessWithTabstopsAfterIt(_VimTest):
+    snippets = ("test", "${1|red,gray|} is ${2:color}\nline 2")
+    keys = "test" + EX + "2"
+    wanted = "gray is color\nline 2"
+
+
+class Choices_MoreThan9Candidates_ShouldWaitForInputs(_VimTest):
+    snippets = ("test", "${1|a,b,c,d,e,f,g,h,i,j,k,l,m,n|} is ${2:a char}")
+    keys = "test" + EX + "1"
+    wanted = "1 is a char"
+
+
+class Choices_EmptyChoiceWillBeDiscarded(_VimTest):
+    snippets = ("test", "${1|a,,c|}")
+    keys = "test" + EX
+    wanted = "1.a|2.c"
+
+
+class Choices_WillNotExpand_If_ChoiceListIsEmpty(_VimTest):
```

Comment:
Add a few more tests for the following cases:

- Mirror:`${1|foo,blah|} - $1`
- Non-latin characters: `${1|öääö|trna|some korean?|}`
- Choices as zero tabstop: `${0|foo,blah|}
- Choices as nested element: `${1:${2|foo,blah|}}` - test both overwriting the default text as well as tabbing through to it.

Reply:
I have a question and a problem with these tests.

* Choices as zero tabstop: `${0|foo,blah|}`, what is the expected behavior ? In my test,   `_update()` is called, but `_do_edit()` is not.

  1. So with input `test<tab>`, the result is `1.foo|2.blah`.
  2. So with input `test<tab>1`, the result is `1` .

  If that is correct , should the test named `Choices_AsZeroTabstop_SelectionWillNotWork` ? Emm, looks like a bug.
  
  
* Mirror: `${1|foo,blah|} - $1` , with input `test<tab>1`, the result is `foo - `. Here I see there is indeed a bug, I will need some time to fix this.

Reply:
a) Yes, that is a nasty corner case I have not considered. UltiSnips exits a snippet as soon as `$0` is reached and will not consider any further transformations inside the snippet. In our case this is not expected behavior, because users will expect that the last placeholder still lets them select.... Since the implementation is fairly difficult, I suggest to just state in the documentation that selections are not supported on `$0` and to throw an error if people try it (add a test for the thrown error, please).

b) cool, great that we found a bug through a test i proposed :).